File: testers/avro/java-quarkus/generated-and-checked-in/com/example/service/UserServiceServer.java
package com.example.service;

import com.example.service.GetUserResponse.Error;
import com.example.service.GetUserResponse.Success;
import com.example.service.Result.Err;
import com.example.service.Result.Ok;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.eclipse.microprofile.reactive.messaging.Incoming;
import org.eclipse.microprofile.reactive.messaging.Outgoing;

/** Kafka RPC server for UserService */
@ApplicationScoped
public record UserServiceServer(UserServiceHandler handler) {
  @Inject
  public UserServiceServer {}

  public UserServiceServer withHandler(UserServiceHandler handler) {
    return new UserServiceServer(handler);
  }

  public CreateUserResponse handleCreateUser(CreateUserRequest request) {
    var result = handler.createUser(request.email(), request.name());
    return switch (result) {
      case Ok ok ->
          new com.example.service.CreateUserResponse.Success(
              request.correlationId(), ((User) ok.value()));
      case Err err ->
          new com.example.service.CreateUserResponse.Error(
              request.correlationId(), ((ValidationError) err.error()));
    };
  }

  public DeleteUserResponse handleDeleteUser(DeleteUserRequest request) {
    var result = handler.deleteUser(request.userId());
    return switch (result) {
      case Ok ok ->
          new com.example.service.DeleteUserResponse.Success(
              request.correlationId(), ((Void) ok.value()));
      case Err err ->
          new com.example.service.DeleteUserResponse.Error(
              request.correlationId(), ((UserNotFoundError) err.error()));
    };
  }

  public GetUserResponse handleGetUser(GetUserRequest request) {
    var result = handler.getUser(request.userId());
    return switch (result) {
      case Ok ok -> new Success(request.correlationId(), ((User) ok.value()));
      case Err err -> new Error(request.correlationId(), ((UserNotFoundError) err.error()));
    };
  }

  public void handleNotifyUser(NotifyUserRequest request) {
    handler.notifyUser(request.userId(), request.message());
  }

  /** Dispatch incoming requests to handler methods */
  @Incoming("user-service-requests")
  @Outgoing("user-service-replies")
  public Object handleRequest(UserServiceRequest request) {
    return switch (request) {
      case GetUserRequest r -> handleGetUser(r);
      case CreateUserRequest r -> handleCreateUser(r);
      case DeleteUserRequest r -> handleDeleteUser(r);
      case NotifyUserRequest r -> {
        handleNotifyUser(r);
        yield null;
      }
    };
  }
}

File: testers/avro/kotlin-quarkus-mutiny/generated-and-checked-in/com/example/service/UserServiceServer.kt
package com.example.service

import com.example.service.GetUserResponse.Error
import com.example.service.GetUserResponse.Success
import com.example.service.Result.Err
import com.example.service.Result.Ok
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import org.eclipse.microprofile.reactive.messaging.Incoming
import org.eclipse.microprofile.reactive.messaging.Outgoing

@ApplicationScoped
/** Kafka RPC server for UserService */
data class UserServiceServer @Inject constructor(val handler: UserServiceHandler) {
  fun handleCreateUser(request: CreateUserRequest): CreateUserResponse {
    val result = handler.createUser(request.email, request.name)
    return when (val __r = result) {
      is Ok<*, *> -> { val ok = __r as Ok<*, *>; com.example.service.CreateUserResponse.Success(request.correlationId, (ok.value as User)) }
      is Err<*, *> -> { val err = __r as Err<*, *>; com.example.service.CreateUserResponse.Error(request.correlationId, (err.error as ValidationError)) }
      else -> throw IllegalStateException("Unreachable")
    }
  }

  fun handleDeleteUser(request: DeleteUserRequest): DeleteUserResponse {
    val result = handler.deleteUser(request.userId)
    return when (val __r = result) {
      is Ok<*, *> -> { val ok = __r as Ok<*, *>; com.example.service.DeleteUserResponse.Success(request.correlationId, (ok.value as Unit)) }
      is Err<*, *> -> { val err = __r as Err<*, *>; com.example.service.DeleteUserResponse.Error(request.correlationId, (err.error as UserNotFoundError)) }
      else -> throw IllegalStateException("Unreachable")
    }
  }

  fun handleGetUser(request: GetUserRequest): GetUserResponse {
    val result = handler.getUser(request.userId)
    return when (val __r = result) {
      is Ok<*, *> -> { val ok = __r as Ok<*, *>; Success(request.correlationId, (ok.value as User)) }
      is Err<*, *> -> { val err = __r as Err<*, *>; Error(request.correlationId, (err.error as UserNotFoundError)) }
      else -> throw IllegalStateException("Unreachable")
    }
  }

  fun handleNotifyUser(request: NotifyUserRequest) {
    handler.notifyUser(request.userId, request.message)
  }

  /** Dispatch incoming requests to handler methods */
  @Incoming("user-service-requests")
  @Outgoing("user-service-replies")
  fun handleRequest(request: UserServiceRequest): Any? {
    return when (val __r = request) {
      is GetUserRequest -> { val r = __r as GetUserRequest; handleGetUser(r) }
      is CreateUserRequest -> { val r = __r as CreateUserRequest; handleCreateUser(r) }
      is DeleteUserRequest -> { val r = __r as DeleteUserRequest; handleDeleteUser(r) }
      is NotifyUserRequest -> {
        val r = __r as NotifyUserRequest
        handleNotifyUser(r)
          null
      }
    }
  }
}